### ISSUE TITLE
Microsoft.Net.ToolsetCompilers package 1.0.0-rc3-20150510-01 triggers phantom code contracts assert

### ISSUE THREAD

**Issue by codespare** [dotnet/roslyn#2908]
In the following method, the second contract requires asserts while the debugger shows the 'key' parameter is not null.
That behavior appeared after updating the toolset compiler package from rc2 to rc3. It is distinct from issue #21 that involved yield statements which are absent in this case.
```c#
public Nesting(ImmutableArray<Nesting<K, V>> prefix, K key, IEnumerable<ImmutableArray<K>> suffixes, IEnumerable<V> values)
{
    Contract.Requires<ArgumentNullException>(prefix != null, nameof(prefix));
    Contract.Requires<ArgumentNullException>(key != null, nameof(key));
    Contract.Requires<ArgumentNullException>(suffixes != null, nameof(suffixes));
    Contract.Requires<ArgumentNullException>(values != null, nameof(values));
    Contract.Requires<ArgumentException>(suffixes.Count() == values.Count(), "Error: nested keys and values collections must have the same count.");// (arguments have \{suffixes.Count()} keys, \{values.Count()} values)");

    _key = key;
    _prefix = prefix;
    _path = new Lazy<ImmutableArray<K>>(
        () => _prefix
            .Select(n => n.Key)
            .Concat(new[] { key })
            .ToImmutableArray(),
        isThreadSafe : true
        );
```
(and yes, that select/concat horror is in need of refactoring :)

**Comment by codespare** [comment]
Hello, I posted an attempt at isolating the issue at https://gist.github.com/codespare/3d8f27231ddcb61eba97
The first file is the source, the second file part of the compiled source decompiled with ILSpy.
The issue doesn't occur in that version, but I found a difference with the decompiled version of the source with issue.
Where the decompiled source without issue goes:
		public TypeDummy(List<TypeDummy<K, V>> prefix, K key, IEnumerable<List<K>> suffixes, IEnumerable<V> values)
		{
			TypeDummy<K, V> <>4__this = this;
			Contract.Requires<ArgumentNullException>(prefix != null, "prefix");
the decompiled source exhibiting the issue is missing the <>n__this = this hoisting statement.
The assert then comes from a nested constructor call.

**Comment by codespare** [comment]
As it turns out, looking at the IL for the example above, ILSpy decompilation is not very fidel. https://gist.github.com/codespare/294f22f195246ad79c3b has the IL for the source compiled with code contract runtime contract checking set to full, which asserts with 'Fail: Precondition failed: key != null  key', and the IL for the source compiled with code contract runtime checking set to none, that passes, including a Trace.Assert(key != null). In the first case, the ldargs indices seem odd, but I cannot really tell how Roslyn codegen can impact CodeContracts rewriting that way.

**Comment by jaredpar** [comment]
@codespare thanks for taking the time to report the issue and get a minimal repro for us.  I took a look and this doesn't appear to be an issue with the Roslyn compiler but instead with Code Contracts.  

To verify this I created the following minimal repro which uses a hand coded version of `Contract.Requires` and added output if any of the contracts failed.  

> https://gist.github.com/jaredpar/a9b27a4d3a7c8c89afa7

Running this using the latest Roslyn compiler verifies that none of the conditions evaluate to `false`.  If this is reproducing locally it would seem to be an issue with the Code Contracts ILRewriter tool.  It has a couple of [known issues](https://github.com/Microsoft/CodeContracts/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen++roslyn) with Roslyn.
